Framework/Algorithms/src/ParallaxCorrection.cpp

```diff
@@ -0,0 +1,188 @@
+  auto lengthValidator =
+      boost::make_shared<Kernel::ArrayLengthValidator<std::string>>();
+  lengthValidator->setLengthMin(1);
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input, validator),
+      "An input workspace.");
+  declareProperty(
+      Kernel::make_unique<Kernel::ArrayProperty<std::string>>("ComponentNames",
+                                                              lengthValidator),
+      "List of instrument components to perform the corrections for.");
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "OutputWorkspace", "", Kernel::Direction::Output),
+      "An output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void ParallaxCorrection::exec() {
+  API::MatrixWorkspace_const_sptr inputWorkspace =
+      getProperty("InputWorkspace");
+  API::MatrixWorkspace_sptr outputWorkspace = getProperty("OutputWorkspace");
+  if (inputWorkspace != outputWorkspace) {
+    outputWorkspace = inputWorkspace->clone();
+  }
+  const std::vector<std::string> componentNames = getProperty("ComponentNames");
+  const auto &instrument = inputWorkspace->getInstrument();
+  const auto &detectorInfo = inputWorkspace->detectorInfo();
+  auto progress =
+      Kernel::make_unique<API::Progress>(this, 0., 1., componentNames.size());
+  for (const auto &componentName : componentNames) {
+    progress->report("Performing parallax correction for component " +
+                     componentName);
+    const auto component = instrument->getComponentByName(componentName);
+    if (!component) {
+      g_log.error() << "No component defined with name " << componentName
+                    << "\n";
+      continue;
+    }
+    if (!component->hasParameter(PARALLAX_PARAMETER) ||
+        !component->hasParameter(DIRECTION_PARAMETER)) {
+      g_log.error() << "No parallax correction defined in IPF for component "
+                    << componentName << "\n";
+      continue;
+    }
+    std::vector<detid_t> detIDs;
+    std::vector<Geometry::IDetector_const_sptr> dets;
+    instrument->getDetectorsInBank(dets, componentName);
+    if (dets.empty()) {
+      const auto det =
+          boost::dynamic_pointer_cast<const Geometry::IDetector>(component);
+      if (!det) {
+        g_log.error() << "No detectors found in component " << componentName
+                      << "\n";
+        continue;
+      }
+      dets.emplace_back(det);
+    }
+    if (!dets.empty()) {
+      detIDs.reserve(dets.size());
+      for (const auto &det : dets) {
+        detIDs.emplace_back(det->getID());
+      }
+      const auto indices = inputWorkspace->getIndicesFromDetectorIDs(detIDs);
+      const std::string parallax =
+          component->getStringParameter(PARALLAX_PARAMETER)[0];
```

Comment:
Unfortunately `ComponentInfo` does not yet support maps of parameters against indexes. It should be moderately easy to add that support. If you had the time to do that, would be much appreciated, though understand if you could not at this stage.